src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java

```diff
@@ -572,50 +565,80 @@ static String normalizeServer(String server) { /*package scope for testing*/
         return StringUtils.join(normalised, ' ');
     }
 
-    @Restricted(NoExternalUse.class)
-    public WebApplicationContext createApplicationContext(LDAPSecurityRealm realm, boolean usePotentialUserProvidedBinding) {
-        Binding binding = new Binding();
-        binding.setVariable("instance", this);
-        binding.setVariable("realmInstance", realm);
-
-        final Jenkins jenkins = Jenkins.getInstance();
-        if (jenkins == null) {
-            throw new IllegalStateException("Jenkins has not been started, or was already shut down");
+    public static final class ApplicationContext {
+        public final AuthenticationManager authenticationManager;
+        public final LdapUserSearch ldapUserSearch;
+        public final LdapAuthoritiesPopulator ldapAuthoritiesPopulator;
+        ApplicationContext(AuthenticationManager authenticationManager, LdapUserSearch ldapUserSearch, LdapAuthoritiesPopulator ldapAuthoritiesPopulator) {
+            this.authenticationManager = authenticationManager;
+            this.ldapUserSearch = ldapUserSearch;
+            this.ldapAuthoritiesPopulator = ldapAuthoritiesPopulator;
         }
+    }
 
-        BeanBuilder builder = new BeanBuilder(jenkins.pluginManager.uberClassLoader);
-        try {
-            File override = getLdapBindOverrideFile(jenkins);
-            if (usePotentialUserProvidedBinding && override.exists()) {
-                builder.parse(new AutoCloseInputStream(new FileInputStream(override)), binding);
-            } else {
-                if (override.exists()) {
-                    LOGGER.warning("Not loading custom " + SECURITY_REALM_LDAPBIND_GROOVY);
-                }
-                builder.parse(new AutoCloseInputStream(LDAPSecurityRealm.class.getResourceAsStream(SECURITY_REALM_LDAPBIND_GROOVY)), binding);
-            }
+    @Restricted(NoExternalUse.class)
+    public ApplicationContext createApplicationContext(LDAPSecurityRealm realm) {
+        DefaultInitialDirContextFactory initialDirContextFactory = new DefaultInitialDirContextFactory(getLDAPURL());
+        if (getManagerDN() != null) {
+            initialDirContextFactory.setManagerDn(getManagerDN());
+            initialDirContextFactory.setManagerPassword(getManagerPassword());
+        }
+        Map<String, String> vars = new HashMap<>();
+        vars.put(Context.REFERRAL, "follow");
+        vars.put("com.sun.jndi.ldap.connect.timeout", "30000"); // timeout if no connection after 30 seconds
+        vars.put("com.sun.jndi.ldap.read.timeout", "60000"); // timeout if no response after 60 seconds
```

Comment:
It is possible that those that have customized the bean builder has done so only to change these timeouts.
It's possible we need to introduce them as system properties or something.